sitemaps_from_robotstxt.py

```diff
@@ -0,0 +1,117 @@
+import json
+import re
+from typing import Optional
+from urllib.parse import urlparse, urljoin
+
+import validators
+from py4j.protocol import Py4JError
+from pyspark.sql.types import StructType, StructField, StringType, ArrayType
+from warcio.recordloader import ArcWarcRecord
+
+from sparkcc import CCSparkJob
+
+class SitemapExtractorJob(CCSparkJob):
+    """Extract sitemap URLs (http://www.sitemaps.org/) from robots.txt WARC files."""
+
+    name = "SitemapExtractor"
+
+    output_schema = StructType([
+        StructField('sitemap_url', StringType(), True),
+        StructField('hosts', ArrayType(elementType=StringType()), True)
+    ])
+
+    # rb: match on raw bytes so we can defer utf-8 decoding to the `sitemap:` line
+    sitemap_pattern = re.compile(rb'^sitemap:\s*(\S+)', re.I)
+
+    robots_txt_processed = None
+    sitemap_urls_found = None
+    sitemap_url_invalid_encoding = None
+    robots_txt_announcing_sitemap = None
+    robots_txt_with_more_than_50_sitemaps = None
+
+
+    def init_accumulators(self, session):
+        super(SitemapExtractorJob, self).init_accumulators(session)
+
+        sc = session.sparkContext
+        self.robots_txt_processed = sc.accumulator(0)
+        self.sitemap_urls_found = sc.accumulator(0)
+        self.sitemap_url_invalid_encoding = sc.accumulator(0)
+        self.robots_txt_announcing_sitemap = sc.accumulator(0)
+        self.robots_txt_with_more_than_50_sitemaps = sc.accumulator(0)
+
+
+    def process_record(self, record: ArcWarcRecord):
+        """ emit: sitemap_url => [host] """
+        if not self.is_response_record(record):
+            # we're only interested in the HTTP responses
+            return
+
+        self.robots_txt_processed.add(1)
+        # robots_txt url/host are lazily computed when we encounter the first valid sitemap URL
+        robots_txt_url = None
+        robots_txt_host = None
+        n_sitemaps = 0
+
+        data = self.get_payload_stream(record).read()
+        for raw_line in data.splitlines():
+            raw_line = raw_line.strip()
+
+            match = SitemapExtractorJob.sitemap_pattern.match(raw_line)
+            if match:
+                sitemap_url = match.group(1).strip()
+                self.sitemap_urls_found.add(1)
+                try:
+                    sitemap_url = sitemap_url.decode("utf-8", "strict")
+                except UnicodeDecodeError as e:
+                    self.get_logger().warn(f'Invalid encoding of sitemap URL {sitemap_url}: {repr(e)}')
+                    self.sitemap_url_invalid_encoding.add(1)
+                    continue
+
+                if robots_txt_url is None:
+                    # first sitemap found: set base URL and get host from URL
+                    robots_txt_url = record.rec_headers['WARC-Target-URI']
```

Comment:
This is not compatible with FastWARC, should be `self.get_warc_header(record, 'WARC-Target-URI')`